crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -55,13 +55,57 @@ pub(crate) fn highlight(
     };
 
     let mut bindings_shadow_count: FxHashMap<Name, u32> = FxHashMap::default();
-    let mut res = Vec::new();
+    // We use a stack for the DFS traversal below.
+    // When we leave a node, the we use it to flatten the highlighted ranges.
+    let mut res: Vec<Vec<HighlightedRange>> = vec![Vec::new()];
```

Comment:
If `HighlightedRange` implemented `Default` could we use `Default::default()` here?

Reply:
No, the outer `Vec` represents the stack and the inner `Vec` corresponds to `res` before. We need the stack data structure to get the parent-child relationship for the flattening logic